File: backend/lib/digital.rb
class Digital < Product
  def prepare(order=nil)
    send_email_to(order.customer)
    generate_discount_voucher_for(order.customer)
  end

  private

  def send_email_to(customer)
  end

  def generate_discount_voucher_for(customer)
  end
end

File: backend-fabricio/src/main/kotlin/order/Digital.kt
package order

import shipping.NotificationService

class Digital(name: String, price: Double) : Product(name, price) {
    private val voucherDiscount: Double = 10.0

    override fun pay(payment: Payment) {
        OrderPaymentEventContainer.addEvent(DigitalProductPaidEvent(payment.order, voucherDiscount, NotificationService()))
    }
}